Realistic SubSample times

ast0815 opened the pull request:
I'm trying to implement a more realistic way of handling SubSamples, especially adding timestamps so supernova trigger decisions can be made over SubSample borders. The new code splits SubSamples into smaller ones, if they are longer than the set window width (e.g. 1 ms). Hit times in each SubSample are relative to the timestamp of that SubSample, so that has to be taken into account when combining trigger information.

For this to make sense, WCSim needs to store some sort of timing information in the events it simulates, e.g. via WCSim/WCSim#277. The digit times in WCSim are currently stored in single precision floats. so large digit times will lead to a loss of timing resolution. At 20 seconds (i.e. the tail end of a SN simulation) the resolution is 20s / 2**21 = 10 us. Much too coarse even for the NHits trigger. This could be handled by either using double precision digit times or adding an event-wide timestamp to the WCSim simulation itself. I guess switching things to doubles would be easier, but this is a WCSim issue and should be handled there.

This is a WIP and the pull request is just here to enable discussions of this development.

TODO:

- [x] Make triggers aware of the timestamp they are relative to
- [x] Reduce code duplication in SubSamples
- [x] Reduce code duplication in PrepareSubSamples
- [x] Implement checks to speed things up when no changes to samples are needed
- [x] Compare new NHits trigger with WCSim implementation

tdealtry commented on DataModel/SubSample.h:
```diff
@@ -3,27 +3,48 @@
+  SubSample() {};
+  SubSample(std::vector<int> PMTid, std::vector<relative_time_t> time);
+  SubSample(std::vector<int> PMTid, std::vector<relative_time_t> time, std::vector<float> charge);
+  SubSample(std::vector<int> PMTid, std::vector<relative_time_t> time, std::vector<float> charge, timestamp_t timestamp);
```
Do we need 2 versions of this? This will simplify:
```suggestion
  SubSample(std::vector<int> PMTid, std::vector<relative_time_t> time, std::vector<float> charge, timestamp_t timestamp = 0);
```

ast0815 replied:
My thinking was that I want to keep the timestamp left of the hit time, because in general it should be the thing with the larger unit (us instead of ns). Also I have read that there are some instances where it is preferable to use overloading over default values: https://google.github.io/styleguide/cppguide.html#Default_Arguments

I am not wedded to this decision, though.

tdealtry replied:
I'm not sure I understand what "timestamp left of the hit time" means, because in the constructor with both, the timestamp is the last argument (i.e. right-most)
Anyway, instead of default constructor arguments, we could instead do constructors calling constructors in the implementation to achieve the same thing, and to reduce code repetition https://en.wikipedia.org/wiki/C++11#Object_construction_improvement

ast0815 replied:
Ahh, sorry. I did not pay attention and was confusing the constructor and the `TimeDifference` method here. Please ignore the argument order argument.

I tried to use the "constructor calls constructor" method, but my compiler did not understand that. I think the default compiler on CentOS7 does not have full C++11 support. Apparently there is a flag to enable experimental support (from before the standard was standardised), but I was not sure I should be using that (and add the flag to the makefile).

tdealtry replied:
Maybe make a method call Construct() that all constructors call? A decision on C++11/14/17/20 support is probably for another PR

tdealtry commented on DataModel/SubSample.h:
```diff
@@ -3,27 +3,48 @@
-  std::vector<int> m_time_int;
-  std::vector<int> m_charge_int;
+  SubSample() {};
+  SubSample(std::vector<int> PMTid, std::vector<relative_time_t> time);
```
Should be able to set the timestamp in this version too (currently, I believe that no tools are actually using the charge information, although it is filled by WCSimReader)
```suggestion
  SubSample(std::vector<int> PMTid, std::vector<relative_time_t> time, timestamp_t timestamp = 0);
```

ast0815 replied:
Same as above. Was trying to avoid default values and use overloading to keep everything consistent.

tdealtry commented on UserTools/Factory/Factory.cpp:
```diff
@@ -8,7 +8,7 @@ if (tool=="DummyTool") ret=new DummyTool;
 if (tool=="TriggerOutput") ret=new TriggerOutput;
 if (tool=="WCSimASCIReader") ret=new WCSimASCIReader;
-if (tool=="nhits") ret=new nhits;
+if (tool=="NHits") ret=new NHits;
```
My vague idea was that "triggers are lower case, other tools are upper cased" (we're going to end up with 30+ tools, so this may be a nice quick way to find triggers/not). What do you think?

ast0815 replied:
I am not a fan. 😆 

I don't like mixing the semantic (what is this thing doing) and syntactic (what is this thing from a coding point of view) hinting like that (I might be using these word wrong). The coding style convention is supposed to help with the latter. Breaking this to have it help with the former is not a good idea, imho. The former should be achieved by explicit names, e.g. `NHitsTrigger` instead of `NHits`.

In that regard we are not doing very well right now. Some tools are named after what they do (e.g. "DataOut", "PrepareSubSamples"), some are named after what they are (e.g. "ReconRandomiser", "DummyTool"), some are named after what they produce (e.g. "test_vertices"), and some are named after an algorithm or something similar, without really explicitly stating what they are doing (e.g. "dimfit", "BONSAI"). Fixing this (and writing an agreed convention down somewhere) would be preferable to breaking the coding convention to achieve something similar but less verbose.

brichards64 replied:
I have to say i prefer the explicit too so my vote would be for NhitsTrigger

tdealtry replied:
Yeah I agree more distinction than upper vs lower case is better. Here's a list of suggestions for all the tools currently available. Please comment!

| Current name | New name |
| - | - |
| **Standard triggers** | |
| nhits | TriggerNHits |
| pass_all | TriggerPassAll |
| test_vertices | TriggerTestVertices | 
| **SN triggers** | |
| dimfit | SNTriggerDimFit |
| **Event reconstruction** | |
| BONSAI | ReconBONSAI |
| EnergeticBONSAI | ReconEnergeticBONSAI |
| ReconRandomiser | ReconRandomiser |
| **Data input** | |
| WCSimASCIReader | DataReaderWCSimASCII |
| WCSimReader | DataReaderWCSimROOT |
| ReconDataIn | DataReaderRecon |
| **Data formatting** | |
| PrepareSubSamples | PrepareSubSamples |
| ReconFilter | PrepareRecons |
| ReconReset | ? |
| **Data output** | |
| DataOut | DataWriterTrigger |
| ReconDataOut | DataWriterRecon |
| **Others** | |
| CUDA | ? |
| DummyTool | ToolDummy |
| template | ToolTemplate |
| TriggerOutput | deprecated? |

We may also want to change `SubSamples` (in the data model) to something different, e.g. `DigitCollection`

ast0815 replied:
Looks pretty good. Though it is still a mix of "NounDescription" and "VerbSomeThing". If we rename everything, we could make it completely consistent in that regard as well. I think this is slightly outside the scope of this pull request though.

tdealtry replied:
Agree it's for another PR. I guess when someone has improvements to the proposal, we can discuss on slack/a meeting

tdealtry commented on UserTools/PrepareSubSamples/PrepareSubSamples.cpp:
```diff
@@ -0,0 +1,80 @@
+    }
+    StreamToLog(DEBUG3);
+  }
+  m_data->ODSamples = new_samples;
```
This looks identical to the ID part, other than the data member difference IDSamples/ODSamples. Can we make a function instead?

ast0815 replied:
Probably a good idea.

tdealtry commented on UserTools/PrepareSubSamples/PrepareSubSamples.cpp:
```diff
@@ -0,0 +1,80 @@
+    ss << "DEBUG: Splitting ID sample";
+    StreamToLog(DEBUG2);
+    std::vector<SubSample> temp_samples = it->Split(m_sample_width, m_sample_overlap);
+    new_samples.insert(new_samples.end(), temp_samples.begin(), temp_samples.end());
```
Looks like we're doing an `insert` operation, even if we're going to end up without any splits (i.e. the output unchanged). Can we do a check on first/last times in the original SubSample against `m_sample_width` first?

tdealtry replied:
Unless `Split()` also does sorting, ensuring positive digits times too...

ast0815 replied:
Split also makes sure that everything is sorted. I considered whether to add a check for whether splitting is necessary, but decided that it is not worth the trouble. In the end we will end up with the same state, even if it took a few useless extra instructions to get there. Since this code is only really needed for MC input without any real-time requirements, I don't think we need to worry about that.

brichards64 replied:
Sorting will be needed for non MC as well. the detector is segmented into like 80 buffer units and they will reply with their data in the window, so we will get 80 streams which will need to be sorted. Although i tend to thing that should be done by the input tool. Prepare sub samples should just cut out overlapping 1ms bits from a 10ms slice.

ast0815 replied:
So there will be SubSamples coming in that need to be merged? Right now it splits every input SubSample completely independently of the other Samples. I kinda assumed that the input will be sliced in the time domain, but each (input) SubSample contains the information for the full detector.

If there is a possibility of merging SubSamples, do I need to keep track of digits that are part of multiple SubSamples? Right now it just copies the digits in the overlapping region. If you were to naively merge the samples together again, you would duplicate those digits.

ast0815 commented:
I guess it is a good idea to do some comparison plots. But is there a particular reason that WCSim's nhits implementation is the "gold standard" aside from it having been available first?

federiconova commented:
Hi Lukas

I think nhits is the SK trigger and so it's a standard for all HK high 
energy triggers. Anyway, it's the only standard we have right now.

Federico

On 10/03/2020 17:23, Lukas Koch wrote:
>
> I guess it is a good idea to do some comparison plots. But is there a 
> particular reason that WCSim's nhits implementation is the "gold 
> standard" aside from it having been available first?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/HKDAQ/TriggerApplication/pull/30?email_source=notifications&email_token=ADJRPQ7TPIC6Y7ZPUGRRQJLRGZZRVA5CNFSM4LEOHH22YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEOMLOCY#issuecomment-597210891>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ADJRPQ7NMFQXFKJ6UJ6OYADRGZZRVANCNFSM4LEOHH2Q>.
>
-- 

_____________________________________________________________
Federico Nova         office: R1 2.89
phone: 01235 44 6265
email:  federico.nova@stfc.ac.uk
_____________________________________________________________
Particle Physics Department            Rutherford Appleton Laboratory
Harwell, Didcot, Oxford, OX11 0QX, UK
_____________________________________________________________

tdealtry commented:
Additionally, while telling people about the software, to get them to start using it, we've been advertising the fact that WCSim & TriggerApplication give identical results (when the config is setup in such a way that the results should be identical). If this is no longer the case, we should advertise how large the difference is (for a high energy, the difference will be negliable (trigger time change by <= 5 ns), but for low energy (where the poisson noise is important) there could be efficiency changes)

tdealtry commented on DataModel/SubSample.cpp:
```diff
@@ -60,7 +48,17 @@ void SubSample::SortByTime(){
+bool SubSample::IsSortedByTime() const {
+  int size = m_time.size();
+  for (int i=0; i<size-1; ++i){
+    if (m_time[i] > m_time[i+1]){
```
```suggestion
    if (m_time[i] >= m_time[i+1]){
```

ast0815 replied:
This change would also make the check return false when consecutive hits have equal times. I would still consider that sorted though.

tdealtry replied:
Sorry you're right. Ignore the suggestion. Got my logic wrong!

ast0815 commented:
I have implemented a new class `TimeDelta` which can be used anywhere we want to keep track of time differences or timestamps. It should be able to handle both long (10s of years) and short (~ps times without loss of precision. Objects of this class can be either created from a floating point number in ns, or by multiplying the provided "unit instances" with a floating point factor, e.g. `0.6 * TimeDelta::ms`.

I then made the TriggerInfo class use this new class and modified NHits to store the full digit time (SubSample timestamp + relative digit time) in the Trigger output.

ast0815 commented:
Here is a histogram of trigger times generated by the new NHits algorithm:

![NHits-vs-WCSim](https://user-images.githubusercontent.com/5884065/76783594-12754d00-67aa-11ea-99bf-14ee660dced4.png)

The input data was generated with a simple electron gun in WCSim and its NDigits trigger. When being run with a trigger, WCSim moves all output digit times 950 ns relative to the trigger time. So if the results were perfectly identical, all trigger times should be at 950 ns.

Does that count as "good enough"?

tdealtry commented:
Thanks @ast0815 I've just run WCSim with its NDigits trigger, with the default daq.mac/WCSimn.mac values, and 10 e- particle gun events. Here are the trigger times (from Date in the EventHeader):
- 500 MeV: 70 ns
- 50 MeV: 75 ns
- 50 MeV and commenting out `//triggertime -= (int)triggertime % 5;` in `WCSimWCTrigger.cc`: 1 75 ns, 3 76 ns, 6 77 ns
So looks like there's a 70-ish ns shift induced somewhere. Any ideas?

It would also be interesting to see on the same plot, the trigger times of WCSim NDigits, and WCSim NoTrigger + TriggerApp nhits, for the same events*, under the same trigger time truncating

*I don't think WCSim uses random numbers after creating digits before the trigger, so the events in the 2 cases should be the same (rather than the random number generators becoming offset)

ast0815 commented:
I'm not sure I follow. You looked at the trigger times reported by WCSim? Why do you think the trigger times should be 0? I assume that WCSim stores digit times relative to the trigger (plus that weird 950 nd shift), but stores the actual trigger time, no?

I don't know how to get the WCSim trigger time out of the output root file. When I figure that out, I can do a direct comparison of the two trigger times: WCSim NDigits vs. WCSim NoTrigger + TriggerApplication NHits.

tdealtry commented:
> I'm not sure I follow. You looked at the trigger times reported by WCSim?
Yep
> Why do you think the trigger times should be 0? 
The trigger time in WCSim _doesn't_ have the 950 ns offset applied, and it seemed to me that you were saying that WCSim triggers should occur at 950 ns (or 0 ns without the offset). It looked a little odd to me, hence me having a look. But either the .mac files we've used are different, and/or (more likely) I've not fully understood the explanation with your plot

> I don't know how to get the WCSim trigger time out of the output root file. When I figure that out, I can do a direct comparison of the two trigger times: WCSim NDigits vs. WCSim NoTrigger + TriggerApplication NHits.
Thanks! When you've got a `WCSimRootTrigger`, it's `GetHeader()->GetDate()`

ast0815 commented:
I don't know when the trigger would be expected. But what WCSim does, is saving all digit times within the trigger *relative* to the trigger time.  So When we run our NHits trigger on those data slices, it should report a trigger at 0 if it finds the trigger at the same time as WCSim did, no matter what that original time was.

On top of that time relative to the trigger, WCSim adds a constant 950 ns to all digit times. I assume this is to make the digit times positive in the output data. So if the NHits algorithm finds the triggers at the same time as WCSim, they should be at 950 ns.

This relative timing and 950 ns offset is only applied in actual triggers within WCSim. When you tell it to not use any triggers and simply save all digits, it should save them with the original times.

tdealtry commented:
Ahh ok. I see now! You've run WCSim NDigits & TriggerApp NHits concurrently. The plot makes sense now. Thanks

ast0815 commented:
I have made a direct comparison of the trigger times now:

![first-trigger-times](https://user-images.githubusercontent.com/5884065/76990120-5cdc0280-693f-11ea-8639-6643f3d1a069.png)

Everything is in nanoseconds.

For a few events, TriggerApplication actually finds multiple triggers. I only got one from WCSim, so I only compare the first trigger reported by TriggerApplication here. I did not spend any time to figure out whether WCSim actually saves more than one trigger per event.

Here is a histogram of the trigger time differences:

![first-trigger-differences](https://user-images.githubusercontent.com/5884065/76990301-a62c5200-693f-11ea-993f-8c72c2c857ad.png)

Same question as before: Is this "good enough"?

ast0815 commented:
The 1000 events were generated with 50 MeV electron particle guns at the centre of the detector shooting in the x direction. 1000 of those events were triggered at least once with a trigger time between 150 ns and 250 ns by the new algorithm, which means the pre-trigger window (set to 400 ns) comfortably included the true vertex time at 0 for 100% of the events.

federiconova commented:
Hi Lukas, thanks. I think that's the most important check. Maybe we 
should test 5 MeV as well as 50 MeV (that's more like our target: a 
trigger threshold in the 3-5 MeV region; the nhits trigger will not be 
good enough there but we can test how badly it fails).

About the pre-trigger window (-400, 950), of course those number are 
largely arbitrary and would be ok to increase them in order to ensure 
capturing all hits from a given trigger.

Federico

On 19/03/2020 12:22, Lukas Koch wrote:
>
> The 1000 events were generated with 50 MeV electron particle guns at 
> the centre of the detector shooting in the x direction. 1000 of those 
> events were triggered at least once with a trigger time between 150 ns 
> and 250 ns by the new algorithm, which means the pre-trigger window 
> (set to 400 ns) comfortably included the true vertex time at 0 for 
> 100% of the events.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/HKDAQ/TriggerApplication/pull/30#issuecomment-601148322>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ADJRPQ23KKYW4TCKVKMUSB3RIIFATANCNFSM4LEOHH2Q>.
>
-- 

_____________________________________________________________
Federico Nova         office: R1 2.89
phone: 01235 44 6265
email:  federico.nova@stfc.ac.uk
_____________________________________________________________
Particle Physics Department            Rutherford Appleton Laboratory
Harwell, Didcot, Oxford, OX11 0QX, UK
_____________________________________________________________

federiconova commented:
(all is based just on the fact that light takes 400 ns at most to go 
through HK, and some late physics may happen on the + side after the 
primary vertex)

tdealtry commented:
I agree with @federiconova. The (-400, 950) readout window is from SKI - to first order we'll probably want to double it (SK max light travel distance ~220 ns, HK ~400 ns)

ast0815 commented:
Since the GPU-version of NHits is outside the scope of what I am trying to achieve, I think this PR is now feature complete and could be merged. Or at least as soon as the build no longer fails because of the container OS version.

tdealtry commented on CONTRIBUTING.md:
```diff
@@ -3,9 +3,12 @@
 *   Use `snake_case` for variables
 *   Add a `m_` prefix to member variables
+*   Add a `s_` prefix to static member (i.e. class-) variables
+*   `ALL-CAPS` enums, implementation constants, and the like, e.g. `s_LONG_TIME`.
```
FYI we're stuck using the WCSim convention for trigger enumerations, at least in some places (handling WCSim files), where the convention is `kUpperCamelCase`

ast0815 replied:
The docs say "try to ... if possible", so I think that one is covered. ;)

brichards64 commented:
Is this intended as an actual PR or a discussion? 

Discussions should go in issues really.

If this is a PR whats its status?

ast0815 commented:
This should be merged.

I think pull requests are very useful also for early discussions of code. I try to add "WIP" to the title as long as it is not yet ready for merging.

brichards64 commented:
Ok iv assigned Tom to review it before i do merge it in as he has had the most experience with it.

can you also resolve the conflicts please.

ast0815 commented:
After pulling the latest changes from "Application" my code no longer compiles:

```
*************** Copying  include/Tool.h ****************
cp ToolDAQ/ToolDAQFramework/src/Tool/Tool.h include/
\n*************** Making  lib/libMyTools.so ****************
cp UserTools/*/*.h include/
cp UserTools/Factory/*.h include/
g++ -g -std=c++11 -Wpedantic -fPIC -shared  UserTools/Factory/Factory.cpp -I include -L lib -lStore -lDataModel -lLogging -o lib/libMyTools.so   -I/opt/ppd/scratch/kdf77245/hk_software/root/include -I/opt/ppd/scratch/kdf77245/hk_software/WCSim/include -L/opt/ppd/scratch/kdf77245/hk_software/root/lib -lCore -lCint -lRIO -lNet -lHist -lGraf -lGraf3d -lGpad -lTree -lRint -lPostscript -lMatrix -lPhysics -lMathCore -lThread -pthread -lm -ldl -rdynamic -L/opt/ppd/scratch/kdf77245/hk_software/WCSim -lWCSimRoot -L ToolDAQ/zeromq-4.0.7/lib -lzmq  -I ToolDAQ/zeromq-4.0.7/include/  -L ToolDAQ/boost_1_66_0/install/lib -lboost_date_time -lboost_serialization -lboost_iostreams -I ToolDAQ/boost_1_66_0/install/include -L/opt/ppd/scratch/kdf77245/hk_software/hk-BONSAI -lWCSimBonsai -I/opt/ppd/scratch/kdf77245/hk_software/hk-BONSAI/bonsai -L -lWCSimEBonsai -I
g++: error: missing path after ‘-I’   
make: *** [lib/libMyTools.so] Error 1 
```

Also I get an error when I source the Setup.sh:

```
$ . Setup.sh                                
-bash: [: =: unary operator expected
Running without energetic BONSAI
```

tdealtry commented:
hmmm maybe workaround to not have to install BONSAI etc doesn't work on CentOS7 (I guess there's new versions of g++, make, etc). When the docker is built, I'll take a look

tdealtry commented:
Ben's put in a fix for the Makefile to stop that error, and I've put in #40 to fix for when we actually do want to use (energetic)BONSAI
Not sure what's going on with your Setup.sh error. It works for me in the CentOS7 docker

brichards64 commented:
fix is merged in so please let us know any build problems you now have

ast0815 commented:
Regarding the Setup file, it complained bout this line: https://github.com/HKDAQ/TriggerApplication/blob/492fdbaca9ea65d8707eeb0b0bc5ada0f99c229d/Setup.sh#L7

It was caused by the empty variable in non-docker environments. I put the variable in quotation marks and it works now.

tdealtry commented on DataModel/SubSample.cpp:
```diff
@@ -1,40 +1,130 @@
+  TimeDelta target_stride = target_width - target_overlap;
+
+  // Ensure everything is sorted
+  assert(IsSortedByTime());
```
```suggestion
  if(!IsSortedByTime())
    SortByTime();
```
@brichards64 can comment, but the idea of the DAQ (and therefore same here in TriggerApp) is to be as good at catching its own errors and carrying on as possible.
Probably needs a comment in CONTRIBUTING alluding to this, and forbidding `assert()`

ast0815 replied:
That would make the method no longer being `const`, since it would potentially change the object. This might be unexpected behaviour when you just expect it to spit out a vector of split SubSamples. Is there maybe another suggested way to "throw an error" or something like that?

brichards64 replied:
Tom, is right ideally we would like to handle errors.

the normal way to throw an error in tooldaq is to return false from from the initialise execute or finalise method that is running, with an associated log statement.

I cant see a situation where you would want to split without the data being sorted, so the sorting proposed by tom is ok.

However another solution would be one could modify your function to have the return type a bool and then the initialise execute or finalise function calling it return false if the function is false. The vector return could then be passed as a referenced argument  variable

brichards64 replied:
Assert, whilst ok for purely software situation, when running on the hardware in the DAQ however, we dont want it taking out a whole TPU. A warning and continuing to operate whilst flagging the error is better. Note that in toolDAQ returning a false form initialise execute or finalise, will inform the framework of the failure allowing it to decide on the action to take. It can stop immediately, try to recover, or close the program graciously, running finalise on each tool rather than a hard stop. This lets you disconnect hardware connections etc.

ast0815 replied:
I made the `Split` method return an empty vector when things are not sorted. Anyone who calls it can then deal with that however they like. I did not touch `PrepareSubSamples`, because it sorts everything beforehand anyway.

tdealtry commented on DataModel/SubSample.h:
```diff
@@ -3,31 +3,46 @@
   void Append(SubSample & sub);
 
-  void Append(std::vector<int> PMTid, std::vector<float> time, std::vector<float> charge);
+  void Append(std::vector<int> PMTid, std::vector<TimeDelta::short_time_t> time, std::vector<float> charge);
```
We probably need `timestamp` as an argument here? (Sorry my PR was merged in first, so falls on you)

ast0815 replied:
I was wondering about that. Where is the `Append` method used and for what?

tdealtry replied:
WCSimReader, when we're reading in a file that's already been triggered (and so is composed of many `WCSimRootTrigger`s

ast0815 replied:
Right. So the hits times have different offsets. That means I'll have to add a timestamp and add the logic to shift the digit time to be relative to the current SubSample.

ast0815 replied:
Done.

tdealtry commented on DataModel/TimeDelta.h:
```diff
@@ -0,0 +1,83 @@
+  /// Member for long time delta
+  long_time_t m_long_time;
+
+  /// Relative unit of long time member, i.e. long_unit / short_unit, both ns so = 1.
```
This contradicts l38 where you say the long time uint is `us`, so this should be `1E6`

ast0815 replied:
It is actually ns, so I need to fix l38.

ast0815 replied:
Done.

tdealtry commented on DataModel/TimeDelta.h:
```diff
@@ -0,0 +1,83 @@
+  static const TimeDelta ns;
+  /// TimeDelta of 1 us
+  static const TimeDelta us;
+  /// TimeDelta of 1 us
```
```suggestion
  /// TimeDelta of 1 ms
```

tdealtry commented on DataModel/TimeDelta.h:
```diff
@@ -0,0 +1,83 @@
+  static const TimeDelta us;
+  /// TimeDelta of 1 us
+  static const TimeDelta ms;
+  /// TimeDelta of 1 us
```
```suggestion
  /// TimeDelta of 1 s
```

tdealtry commented on DataModel/TriggerInfo.h:
```diff
@@ -5,26 +5,40 @@
   TriggerInfo();
 
+  /// Add a trigger, all times in ns
   void AddTrigger(TriggerType_t type, double starttime, double endtime, double triggertime, std::vector<float> info);
```
Any reason why you're keeping the old version with potentially unsafe times?

ast0815 replied:
I kept it, so I would not have to go through all tools and change everything at the same time. I can mark it as deprecated.

tdealtry replied:
Deprecated sounds good

tdealtry commented on DataModel/TriggerInfo.h:
```diff
@@ -5,26 +5,40 @@
+  /// Add a trigger
+  void AddTrigger(TriggerType_t type, TimeDelta starttime, TimeDelta endtime, TimeDelta triggertime, std::vector<float> info);
+
+  /// Add all triggers from aother TriggerInfo object
```
```suggestion
  /// Add all triggers from another TriggerInfo object
```

tdealtry commented on UserTools/PrepareSubSamples/PrepareSubSamples.cpp:
```diff
@@ -0,0 +1,104 @@
+  verbose = 0;
+  m_variables.Get("verbose", verbose);
+
+  ss << "DEBUG: In PrepareSubSamples::Initialise";
+  StreamToLog(DEBUG1);
+
```
```suggestion
```
Don't need lines like this in Initalise/Execute/Finalise - just increase toolchain verbosity to 2 or more

tdealtry commented on UserTools/PrepareSubSamples/PrepareSubSamples.cpp:
```diff
@@ -0,0 +1,104 @@
+  m_sample_width = temp_width * TimeDelta::ns;
+  m_sample_overlap = temp_overlap * TimeDelta::ns;
+
+  ss << "DEBUG: Leaving PrepareSubSamples::Initialise";
+  StreamToLog(DEBUG1);
```
```suggestion
```

tdealtry commented on UserTools/PrepareSubSamples/PrepareSubSamples.cpp:
```diff
@@ -0,0 +1,104 @@
+
+bool PrepareSubSamples::Execute(){
+
+  ss << "DEBUG: Entering PrepareSubSamples::Execute";
+  StreamToLog(DEBUG1);
```
```suggestion
```

tdealtry commented on UserTools/PrepareSubSamples/PrepareSubSamples.cpp:
```diff
@@ -0,0 +1,104 @@
+    m_data->ODSamples = SplitSubSampleVector(m_data->ODSamples);
+  }
+
+  ss << "DEBUG: Exiting PrepareSubSamples::Execute";
+  StreamToLog(DEBUG1);
```
```suggestion
```

tdealtry commented on UserTools/PrepareSubSamples/PrepareSubSamples.h:
```diff
@@ -0,0 +1,51 @@
+  /// The desired SubSample overlap time
+  TimeDelta m_sample_overlap;
+
+  int verbose;
+
+  std::stringstream ss;
```
```suggestion
  int m_verbose;

  std::stringstream m_ss;
```
At least I've changed the template tool, and some others, since the coding conventions were set...

tdealtry commented on UserTools/PrepareSubSamples/README.md:
```diff
@@ -0,0 +1,29 @@
+
+## Configuration
+
+Describe any configuration variables for PrepareSubSamples.
```
```suggestion
```

ast0815 commented:
I did a timing test using the default WCSim output used for WCSimReaderTest.

Here is the result of the old nhits algorithm:

```
[2]: nhits::Execute() run timing stats:
 CPU Time (s): Total = 0.01
             Average = 0.001 over 10 runs
  Range of 0 to 0.01
 Real Time (s): Total = 0.0181842
              Average = 0.00181842 over 10 runs
  Range of 0.00144815 to 0.0021441
```

And here is the result of the new one:

```
[2]: Finalising nhits1
[2]: nhits::Execute() run timing stats:
 CPU Time (s): Total = 0
             Average = 0 over 10 runs
  Range of 0 to 0
 Real Time (s): Total = 0.00524998
              Average = 0.000524998 over 10 runs
  Range of 0.000472784 to 0.000741005
```

It looks like the new algorithm is at least a factor 3~4 faster than the old one. I suspect that factor would be quite a bit larger for events later in the subsample window though.

tdealtry commented:
That's great! Thanks @ast0815!
Ben, over to you now. I think the only thing pending is whether the `assert` statement in `DataModel/SubSample.cpp` is allowed, and if not how the error handling should be done

brichards64 commented:
> That's great! Thanks @ast0815!
> Ben, over to you now. I think the only thing pending is whether the `assert` statement in `DataModel/SubSample.cpp` is allowed, and if not how the error handling should be done

commented on the code and thrown i back for changing

brichards64 commented on DataModel/SubSample.cpp:
```diff
@@ -1,43 +1,140 @@
-SubSample::SubSample(std::vector<int> PMTid, std::vector<float> time, std::vector<float> charge)
-{
+  // Check that all vectors are same size
   assert(PMTid.size() == time.size() && PMTid.size() == charge.size());
```
another assert here

brichards64 replied:
also open { should be inline

brichards64 commented on DataModel/SubSample.cpp:
```diff
@@ -1,43 +1,140 @@
-void SubSample::Append(std::vector<int> PMTid, std::vector<float> time, std::vector<float> charge)
+void SubSample::Append(const std::vector<int> PMTid, const std::vector<TimeDelta::short_time_t> time, const std::vector<float> charge, const TimeDelta timestamp)
 {
   assert(PMTid.size() == time.size() && PMTid.size() == charge.size());
```
another assert here

brichards64 replied:
could use the return here to be bool rather than void

brichards64 replied:
also open { should be inline for all methods

ast0815 replied:
These were there before I even touched the code... 🙄

_Sigh_. On it.

ast0815 replied:
Done.

ast0815 commented:
Done.
